clippy_lints/src/methods/option_take_on_temporary.rs

```diff
@@ -0,0 +1,33 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_errors::Applicability;
+use rustc_hir::Expr;
+use rustc_lint::LateContext;
+use rustc_span::sym;
+
+use super::OPTION_TAKE_ON_TEMPORARY;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>, recv: &'tcx Expr<'_>) {
+    // Checks if expression type is equal to sym::Option and if the expr is not a syntactic place
+    if is_expr_option(cx, expr) & !expr.is_syntactic_place_expr() {
```

Comment:
So I tried checking recv, instead, I am not sure if that was what I was meant to do.
This does not seem to have really helped;
Sorry I didn't really understand everything you said here and in the conversation above

Reply:
Perhaps try and check if `recv` is another `ExprKind::MethodCall(path, ..)` where path is `core::option::Option::as_ref`. That would catch this particular pattern.

Reply:
> So I tried checking recv, instead, I am not sure if that was what I was meant to do. This does not seem to have really helped; Sorry I didn't really understand everything you said here and in the conversation above

Yeah, with `recv` instead of `expr` is what I mean, in both places

Reply:
> Perhaps try and check if `recv` is another `ExprKind::MethodCall(path, ..)` where path is `core::option::Option::as_ref`. That would catch this particular pattern.

I don't understand how the other lints do this, do you have any simple example ?

Reply:
I tried looking at https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.ExprKind.html#variant.MethodCall, but it's an enum, it does not return anything,
I am extremely confused

Reply:
Ok, so I guess there are two ways: 1. Look into `methods/mod.rs`, it has functionality to simply match chains of method calls or 2. Check the type of `recv` which because we've matched on `Option::take` must be an Option, use .types().next() on its GenericArgs and check if that type is a Ref.

The second way had the benefit of not being fooled by using intermediate bindings.

Reply:
It would however make for a false positive if we use `let mut x = Some("foo"); if blargh { x.take() }` if `x` is observed afterwards. I'm unsure if that sort of code pattern is common enough to warrant using the other approach.

Reply:
So I came up with
```rs
fn has_as_ref_codepath(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
    match expr.kind {
        ExprKind::MethodCall(core::option::Option::as_ref, _, _) => todo!(),
        _ => todo!(),
    }
    true
}

```
only, 
```
error[E0533]: expected unit struct, unit variant or constant, found associated function `core::option::Option::as_ref`
  --> clippy_lints/src/methods/option_take_on_temporary.rs:44:30
   |
44 |         ExprKind::MethodCall(core::option::Option::as_ref, _, _) => todo!(),
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
> Ok, so I guess there are two ways: 1. Look into `methods/mod.rs`, it has functionality to simply match chains of method calls or 2. Check the type of `recv` which because we've matched on `Option::take` must be an Option, use .types().next() on its GenericArgs and check if that type is a Ref.
> 
> The second way had the benefit of not being fooled by using intermediate bindings.

So wait now I need to do this instead of what ? matching against the recv `kind` field? using recv all along?

Reply:
Sorry, I am on mobile, so I tend to be brief, which may sometimes make for unclear communication. Unfortunately, `Option::as_ref` has no symbol we can easily use, so we'll need `clippy_utils::match_def_path` (I think) to match the function path. Again, you may want to look at other method lints.

Reply:
So it would be something like : 
```rs
fn has_expr_as_ref_path(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
    if let Some(ref_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
    match_def_path(cx, ref_id, &paths::ASREF_TRAIT)
}

```
?
and also why can't I use if match in the code ?
```
error: expected `{`, found `;`
  --> clippy_lints/src/methods/option_take_on_temporary.rs:38:81
   |
38 |     if let Some(ref_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
   |     -- this `if` expression has a condition, but no block                       ^ expected `{`
```

Reply:
That doesn't seem to be working,

Reply:
`Option::as_ref` is *not* part of the `AsRef` trait impl (and cannot, because it then would need to return `&Option<&T>` thus returning a ref to a dropped temp value). The right path to compare would be `&["core", "option", "Option", "as_ref"]`.

Reply:
```diff
modified   clippy_lints/src/methods/option_take_on_temporary.rs
@@ -1,4 +1,6 @@
 use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::match_def_path;
+use clippy_utils::paths;
 use clippy_utils::source::snippet_with_applicability;
 use clippy_utils::ty::is_type_diagnostic_item;
 use rustc_errors::Applicability;
@@ -10,7 +12,7 @@ use super::OPTION_TAKE_ON_TEMPORARY;
 
 pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, _expr: &'tcx Expr<'_>, recv: &'tcx Expr<'_>) {
     // Checks if expression type is equal to sym::Option and if the expr is not a syntactic place
-    if is_expr_option(cx, recv) & !recv.is_syntactic_place_expr() {
+    if is_expr_option(cx, recv) & !recv.is_syntactic_place_expr() & has_expr_as_ref_path(cx, recv) {
         let mut applicability = Applicability::MachineApplicable;
         span_lint_and_sugg(
             cx,
@@ -31,3 +33,10 @@ fn is_expr_option(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
     let expr_type = cx.typeck_results().expr_ty(expr);
     is_type_diagnostic_item(cx, expr_type, sym::Option)
 }
+
+fn has_expr_as_ref_path(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+    if let Some(ref_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id) && match_def_path(cx, ref_id, &["core", "option", "Option", "as_ref"]) {
+        return true;
+    }
+    false
+}
```
This doesn't seem right

Reply:
it seems clear to me that the condition didn't pass and the suggestion was not made, causing the changes that occured

Reply:
I also noticed the mistake in `+    if is_expr_option(cx, recv) & !recv.is_syntactic_place_expr() & has_expr_as_ref_path(cx, recv) {`,
changed it to `+    if is_expr_option(cx, recv) && !recv.is_syntactic_place_expr() && has_expr_as_ref_path(cx, recv) {`
this seems to make even more warnings